Don't dereference null when DNSServiceGetAddrInfo fails immediately

achristensen07 opened the pull request:
#### 4b93898a545affdf81989b1df86324c6aff889ac
<pre>
Don&apos;t dereference null when DNSServiceGetAddrInfo fails immediately
<a href="https://bugs.webkit.org/show_bug.cgi?id=257977">https://bugs.webkit.org/show_bug.cgi?id=257977</a>
rdar://110383873

Reviewed by David Kilzer.

Ref::leakRef sets the pointer to null, then we were using that value.
With some code modification I verified that this is what was happening in the crash reports.

* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
(WebCore::DNSResolveQueueCFNet::performDNSLookup):

Canonical link: <a href="https://commits.webkit.org/265086@main">https://commits.webkit.org/265086@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d76603746cd68007aa06e6447d415ec236e5c14b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9780 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10028 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10272 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11432 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/9524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9789 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12013 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9979 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12451 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9934 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10751 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11585 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8056 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/8876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/9154 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/9024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12352 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/9525 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/8694 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/12919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1108 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9297 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
```diff
@@ -165,10 +165,11 @@ void DNSResolveQueueCFNet::performDNSLookup(const String& hostname, Ref<Completi
     ASSERT(isMainThread());
     DNSServiceRef service { nullptr };
-    DNSServiceErrorType result = DNSServiceGetAddrInfo(&service, kDNSServiceFlagsReturnIntermediates, kDNSServiceInterfaceIndexAny, kDNSServiceProtocol_IPv4 | kDNSServiceProtocol_IPv6, hostname.utf8().data(), dnsLookupCallback, &wrapper.leakRef());
+    auto& leakedWrapper = wrapper.leakRef();
```
This is an unusual way to move an rvalue reference, but I guess it works looking at `Ref::leakRef()`.  Would be nice if we could use a plain `WTF::Move()` or `std::exchange()` instead to make it more obvious what's happening.

Won't `leakedWrapper` leak when `result == kDNSServiceErr_NoError`, though?  If so, please fix before landing.

ddkilzer replied:
Or does `DNSServiceGetAddrInfo()` consume `leakedWrapper` in the non-error case?

achristensen07 replied:
dnsLookupCallback re-adopts the wrapper in that case.  This code doesn't change that case.

ddkilzer replied:
Yes, `dnsLookupCallback()` consumes `leakedWrapper` when it's called, apparently in the no-error case.

ddkilzer reviewed (approved):
r=me

achristensen07 committed:
Don't dereference null when DNSServiceGetAddrInfo fails immediately
https://bugs.webkit.org/show_bug.cgi?id=257977
rdar://110383873

Reviewed by David Kilzer.

Ref::leakRef sets the pointer to null, then we were using that value.
With some code modification I verified that this is what was happening in the crash reports.

* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
(WebCore::DNSResolveQueueCFNet::performDNSLookup):

Canonical link: https://commits.webkit.org/265086@main
